control-plane/csi-controller/src/client.rs

```diff
@@ -0,0 +1,373 @@
+    /// obtaining the client instance.
+    pub fn initialize(endpoint: String) -> Result<()> {
+        if REST_CLIENT.get().is_some() {
+            return Err(anyhow!("API client already initialized"));
+        }
+
+        let rest_client = reqwest::Client::builder()
+            .danger_accept_invalid_certs(true)
+            .build()
+            .expect("Failed to build REST client");
+
+        REST_CLIENT.get_or_init(|| Self {
+            base_url: format!("{}/v0", endpoint),
+            rest_client,
+        });
+
+        debug!("API client is initialized with endpoint {}", endpoint);
+        Ok(())
+    }
+
+    /// Obtain client instance. Panics if called before the client
+    /// has been initialized.
+    pub fn get_client() -> &'static MayastorApiClient {
+        REST_CLIENT.get().expect("Rest client is not initialized")
+    }
+}
+
+/// Generate a getter for a given collection URI.
+macro_rules! collection_getter {
+    ($name:ident, $t:ty, $urn:expr) => {
+        pub async fn $name(&self) -> Result<Vec<$t>, ApiClientError> {
+            self.get_collection::<$t>($urn).await
+        }
+    };
+}
+
+impl MayastorApiClient {
+    async fn get_collection_item<R>(&self, urn: UrnType<'_>) -> Result<R, ApiClientError>
+    where
+        for<'a> R: Deserialize<'a>,
+    {
+        let response = self.do_get(&urn).await.map_err(|e| {
+            ApiClientError::ServerCommunicationError(format!(
+                "Failed to get {:?}, error = {}",
+                urn, e
+            ))
+        })?;
+
+        // Check HTTP status code.
+        match response.status() {
+            StatusCode::OK => {}
+            StatusCode::NOT_FOUND => {
+                let (rtype, rname) = urn.classify();
+                return Err(ApiClientError::ResourceNotExists(format!(
+                    "{} {} not found",
+                    rtype, rname
+                )));
+            }
+            http_status => {
+                return Err(ApiClientError::GenericOperationError(format!(
+                    "Failed to GET {:?}, HTTP error = {}",
+                    urn, http_status,
+                )))
+            }
+        };
+
+        // Get response body if request succeeded.
+        let body = response.bytes().await.map_err(|e| {
+            ApiClientError::InvalidResponseError(format!(
+                "Failed to obtain body from HTTP response while getting {}, error = {}",
+                urn, e,
+            ))
+        })?;
+
+        serde_json::from_slice::<R>(&body).map_err(|e| {
+            ApiClientError::InvalidResponseError(format!(
+                "Failed to deserialize object {}, error = {}",
+                std::any::type_name::<R>(),
+                e
+            ))
+        })
+    }
+
+    // Get one resource instance.
+    async fn do_get(&self, urn: &UrnType<'_>) -> Result<Response, Error> {
+        let u = format!("{}/{}", self.base_url, urn);
+        let uri = Url::parse(&u).unwrap();
```

Comment:
Should this `unwrap` be a `?` ?
(same for other similar usages)

Reply:
Good point.
We use URLs which are constructed by us, so parsing should always succeed. But as a defensive logic we could have it.